app/models/spree/order.rb

```diff
@@ -369,7 +369,11 @@ def ship_total
     # Creates new tax charges if there are any applicable rates. If prices already
     # include taxes then price adjustments are created instead.
     def create_tax_charge!
-      Spree::TaxRate.adjust(self)
+      clear_legacy_taxes!
+
+      Spree::TaxRate.adjust(self, line_items)
+      Spree::TaxRate.adjust(self, shipments) if shipments.any?
+      tax_enterprise_fees!
     end
```

Comment:
IMO this might be the right time to extract `#create_tax_charge!` out of the `after_create` callback. Although some `before_*` callbacks might be legit, I tend to find the `after_*` rather code-smelly. This has been discussed in the Rails community in-depth in places like [Ruby science](https://github.com/thoughtbot/ruby-science/blob/47ae712a002c64e0d6ede3e4eb98fcfeaa9a4d39/book/code_smells/callback.md) or [here](https://www.thoughtbot.com/upcase/videos/callbacks).

Moving it to a service and calling it from `OrderWorkflow#after_transition_hook` would be a good starting point IMO. We should call it when `order.state == "payment"` if I'm not mistaken.

Reply:
I've been thinking this as well. I'm not sure if we even need to call it from the `after_save` callback at all. I think most of the time when an order is first created there's no need to tax anything...

Reply:
well, that's even better!